### PR TITLE
Add mode argument to `chainerx.Take` 

### LINKED ISSUE #8195: ChainerX take allows OOB access
I think we should not allow this?
```
>>> x = chx.array([1,2])
>>> y = x.take(chx.array([3]), axis=0)
>>> y
array([2], shape=(1,), dtype=int64, device='native:0')
```
numpy raises an exception for this.

### PR DESCRIPTION
**Pull request by emcastillo**
Aims to fix #8195 

However,
When the indexes is a cuda device array, there is no way to check the bounds without incurring in device-host synchronization.

To discuss
+ Consistency between backends
+ Related to linalg error checking

### CODE REVIEW

File: tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py [hunk 1]
```diff
@@ -254,6 +254,21 @@ def forward_xp(self, inputs, xp):
...
+@pytest.mark.parametrize('shape,indices,axis', [
+    # Invalid: Index out of bounds
+    ((2, 3), [2], 0),
+    ((2, 3), [-3], 0),
+])
+def test_take_index_error(device, shape, indices, axis):
+    a = array_utils.create_dummy_ndarray(chainerx, shape, 'float32')
+    indices = numpy.array(indices).astype(numpy.int32)
+    if device.backend.name == 'cuda':
+        pytest.skip()
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py]: How about passing a skip body something along.
```suggestion
        pytest.skip('take with cuda backend does not raise for index error')
```

Comment (emcastillo) [reply tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py]: Done 👍

**Comment by hvy** [comment]
LGTM! 

Jenkins, test this please.

**Comment by hvy** [comment]
Take was apparently intentionally designed to `wrap`, diverging from NumPy. We have to tackle this differently, e.g. by introducing the `mode` argument.

File: chainerx_cc/chainerx/cuda/cuda_device/indexing.cu [hunk 2]
```diff
@@ -160,9 +160,13 @@ void TakeImpl(Device& device, const Array& a, const Array& indices, int8_t axis,
         int64_t total_size = out_indexer.total_size();
         int64_t grid_size = (total_size + kMaxBlockSize - 1) / kMaxBlockSize;
         int64_t block_size = std::min<TIndex>(total_size, kMaxBlockSize);
-
-        TakeCudaKernel<<<grid_size, block_size>>>(
-                a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
+        if (mode == TakeMode::kWrap) {
+            TakeCudaKernel<<<grid_size, block_size>>>(
+                    a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
+        } else {
```
Comment (emcastillo) [inline chainerx_cc/chainerx/cuda/cuda_device/indexing.cu]: This is still on going, different kernels based on the mode

**Comment by hvy** [comment]
Could you fix the compilation error (reported by Travis)?

File: tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py [hunk 3]
```diff
@@ -236,6 +239,9 @@ def forward_xp(self, inputs, xp):
         axis = self.axis
         indices_type = self.indices_type
         a, = inputs
+        if (xp is chainerx and self.mode == 'raise'
+                and 'cuda' in xp.get_default_device().name):
+            pytest.skip('CUDA is not supportted with mode="raise"')
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py]: A nit but how about aligning the way we skip tests (https://github.com/chainer/chainer/pull/8197/files#diff-d2a5309d1c7b10cc433eb74746d256c3R220)?

File: chainerx_cc/chainerx/python/array.cc [hunk 4]
```diff
@@ -240,25 +240,38 @@ void InitChainerxArrayConversion(pybind11::module& m, py::class_<ArrayBody, Arra
...
+                  tmode = IndexBoundsMode::kDefault;
+              } else if (mode.value() == "raise") {
+                  tmode = IndexBoundsMode::kRaise;
+              } else if (mode.value() == "wrap") {
+                  tmode = IndexBoundsMode::kWrap;
+              } else if (mode.value() == "clip") {
+                  tmode = IndexBoundsMode::kClip;
+              } else {
+                  throw py::value_error{"mode must be 'raise', 'wrap', or 'clip'"};
+              }
```
Comment (hvy) [inline chainerx_cc/chainerx/python/array.cc]: [`.value()` does an existence check](https://github.com/abseil/abseil-cpp/blob/master/absl/types/optional.h#L463-L490), let's just check it once in favor of performance, maybe assign it to a local.

File: chainerx_cc/chainerx/routines/indexing.cc [hunk 5]
```diff
@@ -308,7 +308,8 @@ std::vector<Array> Nonzero(const Array& a) {
     if (count_nonzero > 0) {
         Array addat_indices = Where(is_nonzero, Maximum(Cumsum(is_nonzero) - 1, 0), 0);
         Array indices = Where(is_nonzero, Arange(a.GetTotalSize(), Dtype::kInt64), 0);
-        a.device().backend().CallKernel<AddAtKernel>(raw_index, std::move(addat_indices), 0, std::move(indices), raw_index);
+        a.device().backend().CallKernel<AddAtKernel>(
+             raw_index, std::move(addat_indices), 0, std::move(indices), raw_index, IndexBoundsMode::kDefault);
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/indexing.cc]: Ditto regarding not using default.

File: chainerx/_docs/routines.py [hunk 6]
```diff
@@ -554,6 +554,11 @@ def _docs_indexing():
     During backpropagation, this function propagates the gradient of the
     output array to the input array ``a``.
 
+Note:
+   This function throws `IndexError` as a result of an out of bounds access
+   when using the ``native`` backend. However, in the ``cuda`` backend
+   there are no checks to prevent device synchronization.
+
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply chainerx/_docs/routines.py]: Updated

File: tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py (hunk 3 above)
Comment (emcastillo) [reply tests/chainerx_tests/unit_tests/routines_tests/test_indexing.py]: linalg tests skips in a similar way

File: chainerx_cc/chainerx/cuda/cuda_device/indexing.cu [hunk 7]
```diff
@@ -156,18 +215,24 @@ void TakeImpl(Device& device, const Array& a, const Array& indices, int8_t axis,
 
         // TODO(niboshi): Calculate kMaxBlockSize per device
         std::lock_guard<std::mutex> lock{*cuda_internal::g_mutex};
-        static const int kMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&TakeCudaKernel<T, TIndex>).block_size;
+        static const int kMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&TakeWrapCudaKernel<T, TIndex>).block_size;
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply chainerx_cc/chainerx/cuda/cuda_device/indexing.cu]: Fixed :)

File: chainerx_cc/chainerx/cuda/cuda_device/indexing.cu [hunk 8]
```diff
@@ -156,18 +215,24 @@ void TakeImpl(Device& device, const Array& a, const Array& indices, int8_t axis,
...
-                a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
+        if (mode == IndexBoundsMode::kRaise) {
+            throw BackendError{"Take with mode='raise' is not supported with CUDA backend"};
+        } else if (mode == IndexBoundsMode::kWrap || mode == IndexBoundsMode::kDefault) {
+            TakeWrapCudaKernel<<<grid_size, block_size>>>(
+                    a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
+        } else if (mode == IndexBoundsMode::kClip) {
+            TakeClipCudaKernel<<<grid_size, block_size>>>(
+                    a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
+        }
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply chainerx_cc/chainerx/cuda/cuda_device/indexing.cu]: Fixed

File: chainerx_cc/chainerx/routines/indexing.cc (hunk 5 above)
Comment (emcastillo) [reply chainerx_cc/chainerx/routines/indexing.cc]: Here the default value is different for cuda, and native, doing a check implies to have another place defining the default behavior which can lead to inconsistencies if it is changed.

File: chainerx_cc/chainerx/native/native_device/pool.cc [hunk 9]
```diff
@@ -179,7 +180,8 @@ class NativeMaxPoolGradGradKernel : public MaxPoolGradGradKernel {
         return Take(
                 col.Transpose(GetSwapSpatialDimensionsAxes(kernel_size.size())).Reshape({col.GetTotalSize()}),
                 indices + offset.Reshape(indices.shape()),
-                0);
+                0,
+                IndexBoundsMode::kDefault);
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply chainerx_cc/chainerx/native/native_device/pool.cc]: Fixed

File: chainerx_cc/chainerx/python/array.cc (hunk 4 above)
Comment (emcastillo) [reply chainerx_cc/chainerx/python/array.cc]: Fixed

File: chainerx_cc/chainerx/routines/indexing.cc (hunk 5 above)
Comment (hvy) [reply chainerx_cc/chainerx/routines/indexing.cc]: You're right, let's leave it as is.

File: chainerx_cc/chainerx/python/array.cc [hunk 10]
```diff
@@ -240,25 +240,41 @@ void InitChainerxArrayConversion(pybind11::module& m, py::class_<ArrayBody, Arra
...
               }
               if (py::isinstance<py::sequence>(indices)) {
                   absl::optional<Dtype> dtype = Dtype::kInt64;
-                  return MoveArrayBody(Array{self}.Take(Array{MakeArray(indices, dtype, false, self->device())}, axis.value()));
+                  return MoveArrayBody(Array{self}.Take(Array{MakeArray(indices, dtype, false, self->device())}, axis.value(), tmode));
               }
               if (py::isinstance<py::array>(indices)) {
-                  return MoveArrayBody(
-                          Array{self}.Take(Array{MakeArrayFromNumpyArray(py::cast<py::array>(indices), self->device())}, axis.value()));
+                  return MoveArrayBody( Array{self}.Take(
```
Comment (hvy) [inline chainerx_cc/chainerx/python/array.cc]: Major issue (haven't checked the line length).
```suggestion
                  return MoveArrayBody(Array{self}.Take(
```

File: chainerx_cc/chainerx/cuda/cuda_device/indexing.cu [hunk 11]
```diff
@@ -156,18 +215,33 @@ void TakeImpl(Device& device, const Array& a, const Array& indices, int8_t axis,
...
+                block_size = std::min<TIndex>(total_size, kWrapMaxBlockSize);
+                TakeWrapCudaKernel<<<grid_size, block_size>>>(
+                    a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
+                break;
+            case IndexBoundsMode::kClip:
+                grid_size = (total_size + kClipMaxBlockSize - 1) / kClipMaxBlockSize;
+                block_size = std::min<TIndex>(total_size, kClipMaxBlockSize);
+                TakeClipCudaKernel<<<grid_size, block_size>>>(
+                    a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
+                break;
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/indexing.cu]: Do you need a `default`?

**Comment by hvy** [comment]
Jenkins, test this please.

File: chainerx_cc/chainerx/cuda/cuda_device/indexing.cu (hunk 11 above)
Comment (emcastillo) [reply chainerx_cc/chainerx/cuda/cuda_device/indexing.cu]: I think I don't 🤔

**Comment by emcastillo** [comment]
Fixed linter errors

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

File: chainerx_cc/chainerx/cuda/cuda_device/indexing.cu [hunk 12]
```diff
@@ -217,30 +291,61 @@ void AddAtImpl(Device& device, const Array& a, const Array& indices, int8_t axis
...
+        static const int kClipMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&AddAtClipCudaKernel<T, TIndex>).block_size;
+        int64_t grid_size;
+        int64_t block_size;
+        switch (mode) {
+            case IndexBoundsMode::kRaise:
+                throw BackendError{"Take with mode='raise' is not supported with CUDA backend"};
+            case IndexBoundsMode::kDefault:
+            case IndexBoundsMode::kWrap:
+                grid_size = (total_size + kWrapMaxBlockSize - 1) / kWrapMaxBlockSize;
+                block_size = std::min<int64_t>(total_size, kClipMaxBlockSize);
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/indexing.cu]: Ops, a careless.
```suggestion
                block_size = std::min<int64_t>(total_size, kWrapMaxBlockSize);
```

File: chainerx_cc/chainerx/python/routines.cc [hunk 13]
```diff
@@ -315,26 +315,39 @@ void InitChainerxEvaluation(pybind11::module& m) {
 void InitChainerxIndexing(pybind11::module& m) {
     // indexing routines
     m.def("take",
-          [](const ArrayBodyPtr& a, py::handle indices, const absl::optional<int8_t>& axis) {
-              if (!axis.has_value()) {
-                  throw NotImplementedError{"axis=None is not yet supported for chainerx.take."};
```
Comment (hvy) [inline chainerx_cc/chainerx/python/routines.cc]: We must keep this check (it should be tested but that's another story)?

Comment (emcastillo) [reply chainerx_cc/chainerx/python/routines.cc]: It is now checked in the cuda implementation

https://github.com/chainer/chainer/pull/8197/files#diff-effbca5a87c4ee8ecee4fc28fd587b8dR225

I wanted to abstract specific implementation details from the binding code

Comment (emcastillo) [reply chainerx_cc/chainerx/python/routines.cc]: Yes, I dont know why I deleted it

**Comment by hvy** [comment]
LGTM! Let's just check the CI.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Pull request merged**